File: lib/features/library/data/import_service.dart
import 'dart:convert';
import 'dart:io';

import 'package:crypto/crypto.dart';
import 'package:drift/drift.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/db/database.dart';
import '../../../core/db/providers.dart';
import '../../../core/scanner/scan_models.dart';
import '../../../core/subtitle/subtitle_parser.dart';

class ImportSummary {
  const ImportSummary({
    required this.worksInserted,
    required this.worksUpdated,
    required this.tracksTotal,
    this.workIds = const {},
    this.trackIds = const {},
    this.scannedRootPath = '',
    this.filesScanned = 0,
    this.unrecognizedDirs = const [],
    this.scanErrors = const [],
  });

  final int worksInserted;
  final int worksUpdated;
  final int tracksTotal;
  final Set<String> workIds;
  final Set<String> trackIds;

  // Debug fields — surfaced in the import-complete dialog so we can see
  // why a scan returned zero works on device.
  final String scannedRootPath;
  final int filesScanned;
  final List<String> unrecognizedDirs;
  final List<String> scanErrors;
}

class ImportService {
  ImportService(this._db);
  final TonariDatabase _db;

  /// Writes a [ScanResult] into the database. Idempotent on re-run with
  /// the same input. Re-scan picks up new tracks, prunes deleted tracks
  /// (per-work scope), and preserves lastPositionMs / playCount on tracks
  /// that are still present.
  ///
  /// [sourceFolderId] links each Work to the [ImportedFolder] it came from
  /// so the player can resolve the corresponding security-scoped bookmark
  /// later.
  Future<ImportSummary> applyScanResult(
    ScanResult scan, {
    String? sourceFolderId,
    Map<String, List<int>> remoteSubtitleBytes = const {},
  }) async {
    var worksInserted = 0;
    var worksUpdated = 0;
    var tracksTotal = 0;
    final workIds = <String>{};
    final trackIds = <String>{};
    final now = DateTime.now();

    // Read + parse subtitles outside the DB transaction — file I/O can be slow
    // and we don't want to hold the SQLite write lock while doing it.
    final parsedSubs = _readAndParseSubtitles(scan, remoteSubtitleBytes);

    await _db.transaction(() async {
      for (final w in scan.works) {
        workIds.add(w.productId);
        final existing = await (_db.select(
          _db.works,
        )..where((row) => row.productId.equals(w.productId))).getSingleOrNull();

        if (existing == null) {
          await _db
              .into(_db.works)
              .insert(
                WorksCompanion.insert(
                  productId: w.productId,
                  title: w.productId,
                  localFolderPath: w.rootPath,
                  localImportedAt: now,
                  createdAt: now,
                  updatedAt: now,
                  importedFolderId: Value(sourceFolderId),
                ),
              );
          worksInserted++;
        } else {
          await (_db.update(
            _db.works,
          )..where((row) => row.productId.equals(w.productId))).write(
            WorksCompanion(
              localFolderPath: Value(w.rootPath),
              updatedAt: Value(now),
              importedFolderId: Value(sourceFolderId),
              needsRescan: const Value(false),
            ),
          );
          worksUpdated++;
        }

        final scannedIds = <String>{};

        for (final a in w.audios) {
          final baseName = _stripExt(a.fileName);
          final tid = trackIdFor(w.productId, a.relativePath);
          scannedIds.add(tid);
          trackIds.add(tid);

          final existingTrack = await (_db.select(
            _db.tracks,
          )..where((row) => row.id.equals(tid))).getSingleOrNull();

          if (existingTrack == null) {
            await _db
                .into(_db.tracks)
                .insert(
                  TracksCompanion.insert(
                    id: tid,
                    workId: w.productId,
                    filePath: a.path,
                    relativePath: Value(a.relativePath),
                    fileName: a.fileName,
                    fileFormat: a.format,
                    fileSizeBytes: a.sizeBytes,
                    durationMs: 0,
                    parentDirName: a.parentDirName,
                    title: baseName,
                    categoryHint: Value(a.categoryHint),
                    createdAt: now,
                    updatedAt: now,
                  ),
                );
          } else {
            await (_db.update(
              _db.tracks,
            )..where((row) => row.id.equals(tid))).write(
              TracksCompanion(
                filePath: Value(a.path),
                relativePath: Value(a.relativePath),
                fileName: Value(a.fileName),
                fileFormat: Value(a.format),
                fileSizeBytes: Value(a.sizeBytes),
                parentDirName: Value(a.parentDirName),
                title: Value(baseName),
                categoryHint: Value(a.categoryHint),
                updatedAt: Value(now),
              ),
            );
          }
          tracksTotal++;
        }

        if (scannedIds.isEmpty) {
          await (_db.delete(
            _db.tracks,
          )..where((row) => row.workId.equals(w.productId))).go();
        } else {
          await (_db.delete(_db.tracks)..where(
                (row) =>
                    row.workId.equals(w.productId) &
                    row.id.isNotIn(scannedIds.toList()),
              ))
              .go();
        }

        // Subtitles for tracks of this work — id == trackId (one-to-one). Upsert
        // preserves timeOffsetMs and any prior translation cache on a rescan.
        final subsForWork = parsedSubs.where((p) => p.workId == w.productId);
        final subtitleIds = <String>{};
        for (final p in subsForWork) {
          final tid = trackIdFor(w.productId, p.audioRelativePath);
          if (!scannedIds.contains(tid)) continue;
          subtitleIds.add(tid);
          final existing = await (_db.select(_db.subtitles)
                ..where((row) => row.id.equals(tid)))
              .getSingleOrNull();
          if (existing == null) {
            await _db.into(_db.subtitles).insert(
                  SubtitlesCompanion.insert(
                    id: tid,
                    trackId: tid,
                    filePath: p.path,
                    fileFormat: p.format,
                    fileHash: p.hash,
                    originalLinesJson: p.cuesJson,
                    createdAt: now,
                    updatedAt: now,
                  ),
                );
          } else {
            await (_db.update(_db.subtitles)
                  ..where((row) => row.id.equals(tid)))
                .write(
              SubtitlesCompanion(
                filePath: Value(p.path),
                fileFormat: Value(p.format),
                fileHash: Value(p.hash),
                originalLinesJson: Value(p.cuesJson),
                updatedAt: Value(now),
              ),
            );
          }
        }
        // Drop subtitle rows whose audio is gone or whose file disappeared.
        if (subtitleIds.isEmpty) {
          if (scannedIds.isNotEmpty) {
            await (_db.delete(_db.subtitles)..where(
                  (row) => row.trackId.isIn(scannedIds.toList()),
                ))
                .go();
          }
        } else {
          await (_db.delete(_db.subtitles)..where(
                (row) =>
                    row.trackId.isIn(scannedIds.toList()) &
                    row.id.isNotIn(subtitleIds.toList()),
              ))
              .go();
        }

        // Non-audio files (image / subtitle / text / other) live in their own
        // table so the on-disk tree can be rendered without intermixing them
        // with playable tracks.
        final fileIds = <String>{};
        Future<void> upsertFile({
          required String relativePath,
          required String filePath,
          required String fileName,
          required String kind,
          required int sizeBytes,
        }) async {
          final fid = workFileIdFor(w.productId, relativePath);
          fileIds.add(fid);
          final existingFile = await (_db.select(
            _db.workFiles,
          )..where((row) => row.id.equals(fid))).getSingleOrNull();
          if (existingFile == null) {
            await _db.into(_db.workFiles).insert(
                  WorkFilesCompanion.insert(
                    id: fid,
                    workId: w.productId,
                    filePath: filePath,
                    relativePath: relativePath,
                    fileName: fileName,
                    fileKind: kind,
                    fileSizeBytes: sizeBytes,
                    createdAt: now,
                    updatedAt: now,
                  ),
                );
          } else {
            await (_db.update(
              _db.workFiles,
            )..where((row) => row.id.equals(fid))).write(
              WorkFilesCompanion(
                filePath: Value(filePath),
                fileName: Value(fileName),
                fileKind: Value(kind),
                fileSizeBytes: Value(sizeBytes),
                updatedAt: Value(now),
              ),
            );
          }
        }

        for (final f in w.images) {
          await upsertFile(
            relativePath: f.relativePath,
            filePath: f.path,
            fileName: f.fileName,
            kind: 'image',
            sizeBytes: f.sizeBytes,
          );
        }
        for (final f in w.subtitles) {
          await upsertFile(
            relativePath: f.relativePath,
            filePath: f.path,
            fileName: f.fileName,
            kind: 'subtitle',
            sizeBytes: f.sizeBytes,
          );
        }
        for (final f in w.textNotes) {
          await upsertFile(
            relativePath: f.relativePath,
            filePath: f.path,
            fileName: f.fileName,
            kind: 'text',
            sizeBytes: f.sizeBytes,
          );
        }
        for (final f in w.others) {
          await upsertFile(
            relativePath: f.relativePath,
            filePath: f.path,
            fileName: f.fileName,
            kind: 'other',
            sizeBytes: f.sizeBytes,
          );
        }

        if (fileIds.isEmpty) {
          await (_db.delete(
            _db.workFiles,
          )..where((row) => row.workId.equals(w.productId))).go();
        } else {
          await (_db.delete(_db.workFiles)..where(
                (row) =>
                    row.workId.equals(w.productId) &
                    row.id.isNotIn(fileIds.toList()),
              ))
              .go();
        }
      }
    });

    return ImportSummary(
      worksInserted: worksInserted,
      worksUpdated: worksUpdated,
      tracksTotal: tracksTotal,
      workIds: workIds,
      trackIds: trackIds,
      scannedRootPath: scan.rootPath,
      filesScanned: scan.filesScanned,
      unrecognizedDirs: scan.unrecognizedDirs,
      scanErrors: scan.errors,
    );
  }

  /// Stable across re-scans for the same logical track.
  static String trackIdFor(String workId, String relativePath) {
    return '$workId|${relativePath.toLowerCase()}';
  }

  /// Stable across re-scans for the same logical non-audio file.
  static String workFileIdFor(String workId, String relativePath) {
    return '$workId|${relativePath.toLowerCase()}';
  }

  static String _stripExt(String fileName) {
    final i = fileName.lastIndexOf('.');
    return i < 0 ? fileName : fileName.substring(0, i);
  }

  /// Returns parsed-and-hashed subtitles for every work in [scan], matched
  /// to their corresponding audio. Two naming conventions are accepted:
  ///   - `track.wav.vtt` (DLsite style — subtitle name minus one ext == audio name)
  ///   - `track.srt`     (generic style — both files share a stem)
  /// First match wins. Unreadable or empty subtitles are silently skipped.
  List<_ParsedSubtitle> _readAndParseSubtitles(
    ScanResult scan, [
    Map<String, List<int>> remoteBytes = const {},
  ]) {
    final out = <_ParsedSubtitle>[];
    for (final w in scan.works) {
      if (w.subtitles.isEmpty || w.audios.isEmpty) continue;
      for (final sub in w.subtitles) {
        final stem = _stripExt(sub.fileName);
        final match = w.audios.firstWhere(
          (a) => a.fileName == stem || _stripExt(a.fileName) == stem,
          orElse: () => _noAudio,
        );
        if (identical(match, _noAudio)) continue;

        try {
          final bytes = remoteBytes[sub.path] ?? File(sub.path).readAsBytesSync();
          if (bytes.isEmpty) continue;
          final hash = sha256.convert(bytes).toString();
          final content = utf8.decode(_stripBom(bytes), allowMalformed: true);
          final cues = SubtitleParser.parse(content, sub.format);
          if (cues.isEmpty) continue;
          out.add(_ParsedSubtitle(
            workId: w.productId,
            audioRelativePath: match.relativePath,
            path: sub.path,
            format: sub.format,
            hash: hash,
            cuesJson:
                jsonEncode([for (final c in cues) c.toJson()]),
          ));
        } catch (_) {
          // unreadable / unparseable subtitle — skip, don't abort import
        }
      }
    }
    return out;
  }

  static List<int> _stripBom(List<int> bytes) {
    if (bytes.length >= 3 &&
        bytes[0] == 0xEF &&
        bytes[1] == 0xBB &&
        bytes[2] == 0xBF) {
      return bytes.sublist(3);
    }
    return bytes;
  }

  static const _noAudio = DetectedAudio(
    path: '',
    relativePath: '',
    fileName: '',
    format: '',
    sizeBytes: 0,
    parentDirName: '',
  );
}

class _ParsedSubtitle {
  const _ParsedSubtitle({
    required this.workId,
    required this.audioRelativePath,
    required this.path,
    required this.format,
    required this.hash,
    required this.cuesJson,
  });

  final String workId;
  final String audioRelativePath;
  final String path;
  final String format;
  final String hash;
  final String cuesJson;
}

final importServiceProvider = Provider<ImportService>((ref) {
  return ImportService(ref.watch(databaseProvider));
});

File: native/TonariCore/Sources/TonariCore/Import/ImportService.swift
import CryptoKit
import Foundation
import GRDB

public struct ImportSummary: Sendable {
    public var worksInserted = 0
    public var worksUpdated = 0
    public var tracksTotal = 0
    public var workIds: Set<String> = []
    /// Skipped because their scan was incomplete; the user can retry later.
    public var incompleteWorks: [String] = []
    /// Already in the library and left untouched (`skipExisting`).
    public var worksSkipped = 0

    public var resultText: String {
        var text = "导入完成：新增 \(worksInserted)，已有 \(worksSkipped) 跳过，共 \(tracksTotal) 音轨。封面和元数据后台补全中。"
        if !incompleteWorks.isEmpty {
            text += "\n\(incompleteWorks.count) 个作品扫描失败，已跳过，可稍后重新导入。"
        }
        return text
    }
}

extension AppDatabase {
    /// Writes a scan into the library; idempotent for the same input. A
    /// rescan adds new files, prunes vanished ones per work, and keeps each
    /// track's position and play count.
    ///
    /// - `reviveTombstoned`: an explicit single-work reimport brings back a
    ///   removed work; a folder import never does.
    /// - `skipExisting`: leave works already in the library untouched.
    /// - `remoteSubtitles`: for remote sources, the subtitle bytes that were
    ///   downloaded, keyed by file path; nil reads local files from disk.
    public func applyScanResult(
        _ scan: ScanResult,
        sourceFolderId: String?,
        reviveTombstoned: Bool = false,
        skipExisting: Bool = false,
        remoteSubtitles: [String: Data]? = nil
    ) throws -> ImportSummary {
        var summary = ImportSummary()
        let skip: Set<String> = skipExisting ? try activeWorkIds(among: scan.works.map(\.productId)) : []
        summary.worksSkipped = skip.count + scan.skippedExisting
        // File IO outside the write transaction.
        let subtitles = scan.works.filter { !skip.contains($0.productId) && !$0.incomplete }
            .flatMap { Self.readSubtitles($0, remote: remoteSubtitles) }
        let now = Date.now

        try writer.write { db in
            for scanned in scan.works {
                if scanned.incomplete {
                    summary.incompleteWorks.append(scanned.productId)
                    continue
                }
                if skip.contains(scanned.productId) { continue }
                let id = scanned.productId
                let existing = try Work.fetchOne(db, key: id)
                if let existing, existing.isRemoved, !reviveTombstoned { continue }
                summary.workIds.insert(id)

                if existing == nil {
                    try Work.shell(id, folderPath: scanned.rootPath, folderId: sourceFolderId, at: now).insert(db)
                    summary.worksInserted += 1
                } else {
                    try db.execute(sql: """
                        UPDATE works SET local_folder_path = ?, updated_at = ?, imported_folder_id = ?, needs_rescan = 0,
                        is_removed = CASE WHEN ? THEN 0 ELSE is_removed END WHERE product_id = ?
                        """, arguments: [scanned.rootPath, now.unixSeconds, sourceFolderId, reviveTombstoned, id])
                    summary.worksUpdated += 1
                }

                var trackIds: [String] = []
                for audio in scanned.files(.audio) {
                    let trackId = Self.fileId(id, audio.relativePath)
                    trackIds.append(trackId)
                    let title = (audio.fileName as NSString).deletingPathExtension
                    let hint = audio.categoryHint
                    if try Track.exists(db, key: trackId) {
                        try db.execute(sql: """
                            UPDATE tracks SET file_path = ?, relative_path = ?, file_name = ?, file_format = ?,
                            file_size_bytes = ?, parent_dir_name = ?, title = ?, category_hint = ?, updated_at = ? WHERE id = ?
                            """, arguments: [audio.path, audio.relativePath, audio.fileName, FileKind.ext(audio.fileName),
                                             audio.sizeBytes, audio.parentDirName, title, hint, now.unixSeconds, trackId])
                    } else {
                        try Track(
                            id: trackId, workId: id, filePath: audio.path, relativePath: audio.relativePath,
                            fileName: audio.fileName, fileFormat: FileKind.ext(audio.fileName), fileSizeBytes: audio.sizeBytes,
                            durationMs: 0, sampleRate: nil, bitRate: nil, categoryHint: hint, userCategory: nil,
                            parentDirName: audio.parentDirName, trackNumber: nil, title: title,
                            alternateQualityPathsJson: [:], lastPositionMs: 0, playCount: 0,
                            createdAt: now, updatedAt: now, titleZh: nil
                        ).insert(db)
                    }
                    summary.tracksTotal += 1
                }
                try Track.filter(Column("work_id") == id && !trackIds.contains(Column("id"))).deleteAll(db)

                // A subtitle row shares its track's id; upserting keeps the
                // user's time offset and any translation.
                var subtitleIds: [String] = []
                for parsed in subtitles where parsed.workId == id {
                    let trackId = Self.fileId(id, parsed.audioRelativePath)
                    guard trackIds.contains(trackId) else { continue }
                    subtitleIds.append(trackId)
                    if var row = try Subtitle.fetchOne(db, key: trackId) {
                        row.filePath = parsed.path
                        row.fileFormat = parsed.format
                        row.fileHash = parsed.hash
                        row.originalLinesJson = parsed.lines
                        row.updatedAt = now
                        try row.update(db)
                    } else {
                        try Subtitle(
                            id: trackId, trackId: trackId, filePath: parsed.path, fileFormat: parsed.format,
                            fileHash: parsed.hash, timeOffsetMs: 0, originalLinesJson: parsed.lines,
                            translatedLinesJson: nil, translatedAt: nil, translatedByModel: nil,
                            createdAt: now, updatedAt: now
                        ).insert(db)
                    }
                }
                try Subtitle.filter(trackIds.contains(Column("track_id")) && !subtitleIds.contains(Column("id"))).deleteAll(db)

                // Everything that isn't audio, so the folder tree can be drawn.
                var fileIds: [String] = []
                for file in scanned.files where file.kind != .audio {
                    let fileId = Self.fileId(id, file.relativePath)
                    fileIds.append(fileId)
                    if try WorkFile.exists(db, key: fileId) {
                        try db.execute(sql: """
                            UPDATE work_files SET file_path = ?, file_name = ?, file_kind = ?, file_size_bytes = ?, updated_at = ?
                            WHERE id = ?
                            """, arguments: [file.path, file.fileName, file.kind.rawValue, file.sizeBytes, now.unixSeconds, fileId])
                    } else {
                        try WorkFile(
                            id: fileId, workId: id, filePath: file.path, relativePath: file.relativePath,
                            fileName: file.fileName, fileKind: file.kind.rawValue, fileSizeBytes: file.sizeBytes,
                            createdAt: now, updatedAt: now
                        ).insert(db)
                    }
                }
                try WorkFile.filter(Column("work_id") == id && !fileIds.contains(Column("id"))).deleteAll(db)
            }
        }
        return summary
    }

    /// Non-removed works among `ids`.
    public func activeWorkIds(among ids: [String]) throws -> Set<String> {
        try reader.read { db in
            try Set(String.fetchAll(db, Work.filter(keys: ids).filter(Column("is_removed") == false).select(Column("product_id"))))
        }
    }

    /// Stable across rescans for the same logical file.
    static func fileId(_ workId: String, _ relativePath: String) -> String {
        "\(workId)|\(relativePath.lowercased())"
    }

    private struct ParsedSubtitle {
        let workId: String
        let audioRelativePath: String
        let path: String
        let format: String
        let hash: String
        let lines: [Subtitle.Line]
    }

    /// Subtitles matched to audio in the same folder, named either
    /// `track.wav.vtt` (DLsite style) or `track.srt` (shared stem).
    private static func readSubtitles(_ work: ScannedWork, remote: [String: Data]?) -> [ParsedSubtitle] {
        let audios = work.files(.audio)
        func dir(_ relativePath: String) -> String { (relativePath as NSString).deletingLastPathComponent }
        return work.files(.subtitle).compactMap { subtitle in
            let format = FileKind.ext(subtitle.fileName)
            guard SubtitleParser.supports(format) else { return nil }
            let stem = (subtitle.fileName as NSString).deletingPathExtension
            guard let audio = audios.first(where: {
                dir($0.relativePath) == dir(subtitle.relativePath)
                    && ($0.fileName == stem || ($0.fileName as NSString).deletingPathExtension == stem)
            }) else { return nil }
            let data: Data
            if let remote {
                guard let downloaded = remote[subtitle.path] else { return nil }
                data = downloaded
            } else {
                do {
                    data = try Data(contentsOf: URL(filePath: subtitle.path))
                } catch {
                    DiagnosticLog.shared.write("import", "subtitle_unreadable", ["path": subtitle.path, "error": "\(error)"])
                    return nil
                }
            }
            let body = data.starts(with: [0xEF, 0xBB, 0xBF]) ? data.dropFirst(3) : data[...]
            let lines = SubtitleParser.parse(String(decoding: body, as: UTF8.self), format: format)
            guard !lines.isEmpty else { return nil }
            return ParsedSubtitle(
                workId: work.productId, audioRelativePath: audio.relativePath, path: subtitle.path, format: format,
                hash: SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined(), lines: lines
            )
        }
    }
}

extension Work {
    /// A freshly imported work before its DLsite metadata is fetched.
    static func shell(_ productId: String, folderPath: String, folderId: String?, at now: Date) -> Work {
        Work(
            productId: productId, title: productId, titleRomaji: nil, translatedTitle: nil, originalProductId: nil,
            circleId: nil, circleName: nil, releaseDate: nil, voiceActors: [], illustrators: [], scenarioWriters: [],
            musicians: [], ageRating: nil, workType: nil, workTypeName: nil, fileFormats: [], genresJson: [],
            fileSize: nil, seriesId: nil, seriesName: nil, descriptionHtml: nil, titleZh: nil, descriptionHtmlZh: nil,
            mainImageUrl: nil, sampleImageUrls: [], mainImageLocalPath: nil, sampleImageLocalPaths: [],
            descriptionImageLocalPaths: [], officialPrice: nil, currentPrice: nil, discountRate: nil, rating: nil,
            ratingCount: nil, dlCount: nil, wishlistCount: nil, reviewCount: nil, rankDay: nil, rankWeek: nil,
            rankMonth: nil, supportedLanguages: [], scrapedAt: nil, localImportedAt: now, localFolderPath: folderPath,
            importedFolderId: folderId, lastPlayedAt: nil, lastPlayedTrackId: nil, isFavorite: false, isRemoved: false,
            needsRescan: false, userRating: nil, userTags: [], notes: nil, createdAt: now, updatedAt: now
        )
    }
}
